### ISSUE TITLE
(De)serialization breaks some dialect-specific table metadata

### ISSUE THREAD

**Issue by donnillo** [sqlalchemy/sqlalchemy#10213]
### Describe the bug

I was trying to cache reflected table metadata but got some unexpected results after unpickling
```python
from sqlalchemy import create_engine, MetaData, Table

engine = create_engine(
    f'postgresql+psycopg2://{USER}:{PASS}@{HOST}:{PORT}/{NAME}')
meta = MetaData()

some_table = Table('some_table', meta, autoload_with=engine)

some_table.c.some_jsonb_column.comparator
```
```
<sqlalchemy.dialects.postgresql.json.JSONB.Comparator at 0x7f75ec8ddc80>
```
```python
import pickle

some_unpickled_table = pickle.loads(pickle.dumps(some_table))
some_unpickled_table.c.some_jsonb_column.comparator
```
```
<sqlalchemy.sql.sqltypes.NullType.Comparator at 0x7fae9b69a840>
```
This made me unable to use JSONB-specific operators
```python
some_unpickled_table.c.some_jsonb_column.has_key('some_key')
```
```
AttributeError: Neither 'Column' object nor 'Comparator' object has an attribute 'has_key'
```
I understand that table metadata may not be designed to be pickled. But anyway, could this be fixed?

### SQLAlchemy Version in Use

1.4.49

### DBAPI (i.e. the database driver)

psycopg2

### Database Vendor and Major Version

PostgreSQL 12.7

### Python Version

3.9.10

### Operating system

Linux

### To Reproduce

```python
from sqlalchemy import create_engine, MetaData, Table
import pickle

engine = create_engine(
    f'postgresql+psycopg2://{USER}:{PASS}@{HOST}:{PORT}/{NAME}')
meta = MetaData()

some_table = Table('some_table', meta, autoload_with=engine)
some_unpickled_table = pickle.loads(pickle.dumps(some_table))
some_unpickled_table.c.some_jsonb_column.has_key('some_key')
```

### Error

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File ~/.pyenv/versions/main/lib/python3.9/site-packages/sqlalchemy/sql/elements.py:855, in ColumnElement.__getattr__(self, key)
    854 try:
--> 855     return getattr(self.comparator, key)
    856 except AttributeError as err:

AttributeError: 'Comparator' object has no attribute 'has_key'

The above exception was the direct cause of the following exception:

AttributeError                            Traceback (most recent call last)
Input In [1], in <module>
      8 some_table = Table('some_table', meta, autoload_with=engine)
      9 some_unpickled_table = pickle.loads(pickle.dumps(some_table))
---> 10 some_unpickled_table.c.some_jsonb_column.has_key('some_key')

File ~/.pyenv/versions/main/lib/python3.9/site-packages/sqlalchemy/sql/elements.py:857, in ColumnElement.__getattr__(self, key)
    855     return getattr(self.comparator, key)
    856 except AttributeError as err:
--> 857     util.raise_(
    858         AttributeError(
    859             "Neither %r object nor %r object has an attribute %r"
    860             % (
    861                 type(self).__name__,
    862                 type(self.comparator).__name__,
    863                 key,
    864             )
    865         ),
    866         replace_context=err,
    867     )

File ~/.pyenv/versions/main/lib/python3.9/site-packages/sqlalchemy/util/compat.py:211, in raise_(***failed resolving arguments***)
    208     exception.__cause__ = replace_context
    210 try:
--> 211     raise exception
    212 finally:
    213     # credit to
    214     # https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/
    215     # as the __traceback__ object creates a cycle
    216     del exception, replace_context, from_, with_traceback

AttributeError: Neither 'Column' object nor 'Comparator' object has an attribute 'has_key'

```

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**implement custom setstate to work around implicit type/comparator** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4805

**Comment by zzzeek** [comment]
this can be fixed for 2.0, I will run a backport on 1.4 but if it runs into bigger problems it would be 2.0 only

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_4** branch:

**implement custom setstate to work around implicit type/comparator** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4806
